### ISSUE TITLE
DNS search options are not passed properly for Windows pods

### ISSUE THREAD

**Issue by rajatchopra** [openvswitch/ovn-kubernetes#336]
The kubelet passes DNS options through the CRI to container runtime. The windows docker runtime does not take care of it properly yet.
Other CNI plugins for Windows take care of DNS settings meanwhile. OVN needs to do the same or we end up with a non-functional DNS in Windows pods.

/cc @alinbalutoiu

**Comment by alinbalutoiu** [comment]
@rajatchopra I am having some issues with the DNS and it looks like a bug in Windows to me.
Tested on both Windows Server 1709 and 1803.

Endpoint details:
```
PS C:\> Get-HnsEndpoint

ActivityId                : ac795c78-a4ca-4c67-b701-8efa2cbd783e
CreateProcessingStartTime : 131716346366580288
DNSServerList             : 8.8.8.8
DNSSuffix                 : default.svc.cluster.local,svc.cluster.local,cluster.local,localdomain
GatewayAddress            : 10.0.3.1
ID                        : 0c72d71b-f1dd-4a9a-af19-ef389cac8d33
IPAddress                 : 10.0.3.3
MacAddress                : 0a-00-00-00-00-2f
Name                      : default_nano-1803
Policies                  : {}
PrefixLength              : 24
Resources                 : @{AllocationOrder=2; Allocators=System.Object[]; ID=ac795c78-a4ca-4c67-b701-8efa2cbd783e; PortOperationTime=0; State=1;
                            SwitchOperationTime=0; VfpOperationTime=0; parentId=b8e9eece-47ab-4587-8c62-3d5f36da3db6}
SharedContainers          : {be166d78fb7f9144bc15907560117a7434f3651e51d583fdb97bb379c0ed2a3e}
StartTime                 : 131716346368549498
State                     : 2
Type                      : transparent
Version                   : 30064771074
VirtualNetwork            : 3e8051a9-23ba-4d31-add5-a30ee81632c1
VirtualNetworkName        : 4cb7456c8810cf9362ea177f45b224ed3646349d55f36fc60cfdd6489133b20a
```
It has the DNS 8.8.8.8 set to it.

The infra container ``ipconfig /all``
```
PS C:\> docker exec be166d78fb7f ipconfig /all

Windows IP Configuration

   Host Name . . . . . . . . . . . . : nano-1803
   Primary Dns Suffix  . . . . . . . :
   Node Type . . . . . . . . . . . . : Hybrid
   IP Routing Enabled. . . . . . . . : No
   WINS Proxy Enabled. . . . . . . . : No
   DNS Suffix Search List. . . . . . : default.svc.cluster.local,svc.cluster.local,cluster.local,localdomain

Ethernet adapter vEthernet (default_nano-1803):

   Connection-specific DNS Suffix  . : default.svc.cluster.local,svc.cluster.local,cluster.local,localdomain
   Description . . . . . . . . . . . : Hyper-V Virtual Ethernet Adapter #4
   Physical Address. . . . . . . . . : 0A-00-00-00-00-2F
   DHCP Enabled. . . . . . . . . . . : No
   Autoconfiguration Enabled . . . . : Yes
   Link-local IPv6 Address . . . . . : fe80::c156:d7a:516f:dd5e%21(Preferred)
   IPv4 Address. . . . . . . . . . . : 10.0.3.3(Preferred)
   Subnet Mask . . . . . . . . . . . : 255.255.255.0
   Default Gateway . . . . . . . . . : 10.0.3.1
   DNS Servers . . . . . . . . . . . : 8.8.8.8
   NetBIOS over Tcpip. . . . . . . . : Disabled
```
It has the DNS server set to 8.8.8.8 as expected.

A random container from within the same POD:
```
PS C:\> docker exec 95676efacbe8 ipconfig /all

Windows IP Configuration

   Host Name . . . . . . . . . . . . : nano-1803
   Primary Dns Suffix  . . . . . . . :
   Node Type . . . . . . . . . . . . : Hybrid
   IP Routing Enabled. . . . . . . . : No
   WINS Proxy Enabled. . . . . . . . : No

Ethernet adapter vEthernet (default_nano-1803):

   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : Hyper-V Virtual Ethernet Adapter #4
   Physical Address. . . . . . . . . : 0A-00-00-00-00-2F
   DHCP Enabled. . . . . . . . . . . : No
   Autoconfiguration Enabled . . . . : Yes
   Link-local IPv6 Address . . . . . : fe80::c156:d7a:516f:dd5e%21(Preferred)
   IPv4 Address. . . . . . . . . . . : 10.0.3.3(Preferred)
   Subnet Mask . . . . . . . . . . . : 255.255.255.0
   Default Gateway . . . . . . . . . : 10.0.3.1
   NetBIOS over Tcpip. . . . . . . . : Disabled
```
It doesn't have DNS server set.

Testing on the infra container the DNS:
```
PS C:\> docker exec be166d78fb7f ping google.com

Pinging google.com [216.58.210.14] with 32 bytes of data:
Reply from 216.58.210.14: bytes=32 time=41ms TTL=126
PS C:\>
```
Testing on the random container from within the same POD:
```
PS C:\> docker exec 95676efacbe8 ping google.com
Ping request could not find host google.com. Please check the name and try again.
PS C:\>
```

Looking at the compartments it all seems fine ``ipconfig /allcompartments /all``:
```
==============================================================================
Network Information for Compartment 2
==============================================================================
   Host Name . . . . . . . . . . . . : WIN-H4SB0C6TQ1Q
   Primary Dns Suffix  . . . . . . . :
   Node Type . . . . . . . . . . . . : Hybrid
   IP Routing Enabled. . . . . . . . : No
   WINS Proxy Enabled. . . . . . . . : No
   DNS Suffix Search List. . . . . . : default.svc.cluster.local,svc.cluster.local,cluster.local,localdomain

Ethernet adapter vEthernet (default_nano-1803):

   Connection-specific DNS Suffix  . : default.svc.cluster.local,svc.cluster.local,cluster.local,localdomain
   Description . . . . . . . . . . . : Hyper-V Virtual Ethernet Adapter #4
   Physical Address. . . . . . . . . : 0A-00-00-00-00-2F
   DHCP Enabled. . . . . . . . . . . : No
   Autoconfiguration Enabled . . . . : Yes
   Link-local IPv6 Address . . . . . : fe80::c156:d7a:516f:dd5e%21(Preferred)
   IPv4 Address. . . . . . . . . . . : 10.0.3.3(Preferred)
   Subnet Mask . . . . . . . . . . . : 255.255.255.0
   Default Gateway . . . . . . . . . : 10.0.3.1
   DNS Servers . . . . . . . . . . . : 8.8.8.8
   NetBIOS over Tcpip. . . . . . . . : Disabled
```

This needs further investigation.

**Comment by rajatchopra** [comment]
@alinbalutoiu Can we manually fix the mtu and dns servers inside the interface after it has been created?
All we need is the interface index of the container:

```
cmd /c "netsh.exe interface ipv4 set subinterface $($ifIndex) mtu=1400 store=persistent"
Set-DnsClientServerAddress -InterfaceIndex $ifIndex -ServerAddresses $dnsAddresses
```

What do you think?

**Comment by alinbalutoiu** [comment]
@rajatchopra For the MTU it should be possible, although I couldn't get netsh to work on the container interface but it seemed that Set-NetIPInterface (https://docs.microsoft.com/en-us/powershell/module/nettcpip/set-netipinterface?view=win10-ps) was working just fine and can be executed from the host itself without requiring any access to the container, and it works on the interface alias which is already known. Will update with that soon.

For the DNS it might be a bit tricky, I'll have to check it out, but my guess is that we'll have to set the DNS inside each container. From what I've seen in the azure CNI plugin, besides setting the DNS on the endpoint, they are also creating the HNS network with that DNS set, which is something that we are not doing currently.
In azure CNI they've made the HNS network creation a part of the CNI plugin, I'm trying to figure out what will be the best way to do it in our case. (obviously the best way  for this would be to actually work when you specify the DNS for the endpoint)

**Comment by alinbalutoiu** [comment]
It is a one time thing, we can put it in the instructions. The questions is, what happens if the DNS gets modified? Then you have to re-create the network and this means the connection will also drop while re-creating the network.

**Comment by rajatchopra** [comment]
@alinbalutoiu That is completely fair. If the DNS search string/DNS-server-list gets changed later, that is fine. Most of the clusters will do just fine by pointing to default kubernetes apiserver service itself alongwith one that resolves external DNS.
i.e. pick up one from the default (or 8.8.8.8) and use one more (172.30.0.1) for the in-cluster DNS resolution and most clusters should work just fine. It is undrdstood that if the servers/search-strings need to be changed later, then one has to recreate the whole network (possibly breaking the existing pods).
It is a stop gap measure, but I feel its better than nothing for today. The bug with one container seeing it and another not seeing the DNS server list is definitely something that needs a fix.

Do you mind pasting the correct arguments to docker network create for DNS server list here? We can include it in the ansible scripts, or at least add it as a comment there.

**Comment by alinbalutoiu** [comment]
@rajatchopra sounds fair enough. Haven't tested it yet though, but it is supposed to work.
The additional arguments should be added here: https://github.com/openvswitch/ovn-kubernetes/blob/master/contrib/roles/windows/kubernetes/tasks/setup_docker.yml#L17
``com.docker.network.windowsshim.dnssuffix`` and ``com.docker.network.windowsshim.dnsservers``

**Comment by mihail-velikov** [comment]
Hey guys,

I have had issues with DNS on windows containers since we started our containerization efforts (since January). I have tried to use the com.docker.network.windowsshim.dnssuffix and com.docker.network.windowsshim.dnsservers when creating the network but it did not work at all. (At least on 2016 RTM using default docker packages).

Since then we have been using a workaround in our containers -> add a task on boot to setup the dns. The MTU problem appeared when we migrated from 2016 RTM to 2019 preview and we applied the same fix -> on boot setup the mtu.

Kind Regards,
Mihail Velikov
